In [ ]:
pip install quandl

In [17]:
!pip install quandl
import quandl
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import style
style.use('ggplot')
# IMPORTANT: Replace 'YOUR_QUANDL_API_KEY' with your actual Quandl API key
# You can obtain an API key from Quandl's website after registering.
quandl.ApiConfig.api_key='YOUR_QUANDL_API_KEY'
apple=quandl.get('WIKI/AAPL')
msft=quandl.get('WIKI/MSFT')
df=pd.DataFrame(index=apple.index)
df['AAPL']=apple['Adj. Close']
df['MSFT']=msft['Adj. Close']

QuandlError: (Status 403) Something went wrong. Please try again. If you continue to have problems, please contact us at connect@quandl.com.

In [8]:
class X:
  {

  }


In [9]:
for i in range(10):
  pass
print(i)

9


In [10]:
1+-2

-1

In [13]:
a='ant'
b="bat"
c='camel'
print(a,b,c,sep='"')
#print(a,b,c)

ant"bat"camel


In [20]:
i=5
while i>0:
  i=i//2
  if i%2==0:
    print('breaking now')
    break
else:
    i+=1
    print(i)

breaking now


In [24]:
for i in range(1,3):
  print("*",end="")
else:
  print("*")

***


In [11]:
import sys

# Read one character directly from sys.stdin
# Note: In Colab, sys.stdin.read(1) might not be interactive.
# If you need interactive input, consider using input() and taking the first char.
ch = sys.stdin.read(1)
print(f"The character read from stream is: '{ch}'")

The character read from stream is: ''


In [31]:
if s1:
    print(f"The first character of s1 is: '{s1[0]}' ")
else:
    print("s1 is an empty string, so there is no first character to print.")

s1 is an empty string, so there is no first character to print.


In [14]:
s="**********"
s = s[:2] + s[3:] # Corrected: removes the character at index 2 using slicing
print (s)

*********


In [9]:
c,b,a=1,3,2
print (a)
print(b)
print(c)

#c, b, a = b, a, c
#print (a,b,c)

a, b, c = c, a, b
print(a,b,c)


2
3
1
1 2 3


In [12]:
import random
v1 = random.random()
v2 = random.random()
print(len(random.sample([1,2,3],2)) > 2)
print(v1 == v2)
print(random.choice([1,2,3]) >=1)
print(v1 >= 1)

False
False
True
False


In [15]:
import platform
platform.python_version()

'3.12.13'

In [18]:
import sys
import math
b1=type(dir(math)) is list
b2=type(sys.path) is list
print( b1 and b2)


True


In [4]:
import math
try:
    # Attempt to convert an empty string, which will raise a ValueError
    print(int(''))
except ValueError:
    print("Caught a ValueError: Cannot convert empty string to int")
else:
    print("No exception was raised in the try block.")

Caught a ValueError: Cannot convert empty string to int


In [12]:
try:
    print(int("0"))   # "0" converts cleanly to 0 → no exception
except NameError:
    print("0")
else:
    print(int(""))    # ← this runs

0


ValueError: invalid literal for int() with base 10: ''

In [13]:
#try:
#    print(int("0"))   # "0" converts cleanly to 0 → no exception
#except NameError:
#    print("0")
#else:
#    print(int(""))    # ← this runs


try:
    print(0/0)   # ZeroDivisionError
    print('division of 0/0')
except:
    print(0/1)   # catches it
    print('division of 0/1')
else:
    print(0/2)   # skipped
    print('division of 0/2')





0.0
division of 0/1


In [14]:
try:
    print(math.sqrt(-1))  # raises ValueError
    print('square root of -1')
except:
    print(math.sqrt(0))   # catches it
    print('square root of 0')
else:
    print(math.sqrt(1))   # skipped
    print('square root of 1 in Else Block execution')


0.0
square root of 0


In [15]:
try:

    print(float("1e1"))   # raises ValueError → bad format
except (NameError, SystemError):
    print(float("1a1"))
    print('caught error')
else:
    print(float("1e1"))
    print('in else block')

10.0
10.0
in else block


In [21]:
s='2A'
try:
  n=int(s)
except ValueError:
  n=2
except ArithmeticError:
  n=1
except:
  n=3
print(n)

2


In [22]:
s='2A'
try:
  n=int(s)
except:
  n=3
except ValueError:
  n=2
except ArithmeticError:
  n=1
print(n)

SyntaxError: default 'except:' must be last (1260553739.py, line 4)

In [23]:
my_list=[1,2,3]
try:
  my_list[3]=my_list[2]
except BaseException as error:
  print(error)

list assignment index out of range


In [26]:
print(len('•') == 1)
print(len("""
""") == 0)
print(chr(ord('A') + 1) == 'B')
print(ord("Z") - ord("z") -- ord("0"))

True
False
True
16


In [30]:
the_list="alpha;beta;gamma".split(";")
the_string=''.join(the_list)
print(the_string.isalpha())



True


In [36]:
print(121 + 1 != '!' + 2 * '2')
print('AbC'.lower () < 'AB')
print('1' + '1' + '1' < '1' * 3)
print('3.14' != str(3.1415))

True
False
False
True


In [38]:
print(str(1-1) in '012345£739'[:2])
print('phd' in 'alpha')
print('deb' not in 'abcde' [::-1])
print('True' not in 'False')

True
False
True
True


In [41]:
#print(rfind("python","r"))
print(sorted("python"))
#print("python".sort ())
print("python".index("th"))

['h', 'n', 'o', 'p', 't', 'y']
2


In [44]:
string1=str(1/3)
print(string1)
dummy=''
for character in string1:
  dummy=dummy+character
print(dummy)
print(dummy[-1])


0.3333333333333333
0.3333333333333333
3


In [52]:
class ClassA:
    var = 1
    def __init__(self, prop):
        prop1 = prop2 = prop  # local variables only, not self.prop1!

class ClassB(ClassA):
    def __init__(self, prop):
        prop3 = prop ** 2     # local variable, prop3 = 4
        super().__init__(prop) # calls ClassA.__init__(2)

Object = ClassB(2)
print(str(Object))

print(str(Object) == 'Object')
print(__name__ =='__main__')
print(ClassA. __module__ )
print(ClassA. __module__ == 'ClassA')
print(len(ClassB.__bases__) == 1)

False
True
__main__
False
True


In [4]:
class Class:
    Variable = 0          # Class Variable (shared across all instances)
    def __init__(self):
        self.value = 0    # Instance Variable (unique to each object instance)
object_1=Class()
object_1. Variable +=1
object_2=Class()
object_2.value += 1
print(object_1.Variable)
print(object_2.value)

print(object_1.Variable+object_2.value)


1
1
2


In [9]:
class A:
  pass
class B(A):
  pass
class C(A):
  pass
class D(B):
  pass

class Class_4 (D, A) : pass
class Class_1(C,D): pass
#class Class_3(A,C): pass
#class Class_2(B,D): pass


In [18]:
class Super:
  def make(self):
    pass
  def doit(self):
    return self.make()
class Sub_A(Super):
  def make(self):
    return 1
class Sub_B(Super):
  pass
a=Sub_A()
b=Sub_B()
#print(a.doit()+b.doit())



TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

In [20]:
class A:
  VarA=1
  def __init__(self):
    self.prop_a=1

class B(A):
  VarB=2
  def __init__(self):
    self.prop_a=2
    self.prop_aa=2
class C(B):
  VarA=3
  def __init__(self):
    super().__init__()

obj_a=A()
obj_b=B()
obj_c=C()

print(obj_b.prop_a)
print(obj_b.prop_a == 3)
print(hasattr(obj_b, 'prop_aa'))
print(isinstance(obj_c,A))
print(B.VarA)
print(B. VarA == 3)



2
False
True
True
1
False


In [25]:
class Class:
  var=1
  def __init__(self,value):
    self.prop=value
object=Class(2)
print('var' in object.__dict__)
print('prop' in Class.__dict__)
print(len(object.__dict__)==1)
print( 'var' in Class.__dict__)




False
False
True
True


In [29]:
class MyClass:
  def __init__(self,size):
    self.queue=[i for i in range(size)]
  def get(self):
    return self.queue
  def get_last(self):
    return self.queue[-1]
  def add_new(self):
    self.queue.append(self.get_last() +1)
Object = MyClass(2)
Object.add_new()
print(Object.get())






[0, 1, 2]


In [3]:
#What is the expected behavior of the following code?
x = 8 ** (1/3)
y = 2.0 if x < 2.3 else 3.0
print(y)

#A. it outputs 2.0
#B. it outputs 2. 5
#C. the code is erroneus and it will not execute
#D. it outputs 3.0

2.0


Let's explain the code in `XaWcXFaygcKR`:

*   `x = 8 ** (1/3)`: This line calculates the cube root of 8. In Python, `**` is the exponentiation operator. `1/3` results in `0.333...`. So, `8 ** (1/3)` calculates 8 raised to the power of 1/3, which is 2. The variable `x` will store the float value `2.0`.

*   `y = 2.0 if x < 2.3 else 3.0`: This is a conditional expression (often called a ternary operator). It means:
    *   If the condition `x < 2.3` is `True`, then `y` is assigned the value `2.0`.
    *   Otherwise (if the condition is `False`), `y` is assigned the value `3.0`.
    Since `x` is `2.0`, and `2.0 < 2.3` is `True`, `y` will be assigned `2.0`.

*   `print(y)`: This line prints the final value stored in the variable `y` to the console.

In [8]:
def foo(x,y,z):
  return x(y) - x(z)
print({foo(lambda x: x % 2, 2, 1) })
#A. 1
#B. 0
#C. -1
#D. an exception is raised


{-1}


Let's explain the code in `nmpgxOQSibV1`:

*   `def foo(x,y,z): return x(y) - x(z)`:
    *   This defines a function named `foo` that takes three arguments: `x`, `y`, and `z`.
    *   The key point here is that `x` is expected to be a function itself. Inside `foo`, `x` is called twice: once with `y` as its argument (`x(y)`) and once with `z` as its argument (`x(z)`).
    *   The function then returns the result of `x(y)` minus `x(z)`.

*   `lambda x: x % 2`:
    *   This is an anonymous function (a lambda function) that takes one argument, also named `x` (this `x` is local to the lambda and doesn't conflict with the `x` in `foo`).
    *   It returns the remainder when its input `x` is divided by 2 (the modulo operator `%`). This essentially tells us if a number is even (remainder 0) or odd (remainder 1).

*   `print({foo(lambda x: x % 2, 2, 1)})`:
    *   Here, `foo` is called with three arguments:
        1.  `x`: The lambda function `lambda x: x % 2`.
        2.  `y`: The integer `2`.
        3.  `z`: The integer `1`.
    *   Let's trace `foo`'s execution:
        *   `x(y)` becomes `(lambda x: x % 2)(2)`, which evaluates to `2 % 2 = 0`.
        *   `x(z)` becomes `(lambda x: x % 2)(1)`, which evaluates to `1 % 2 = 1`.
        *   The return value of `foo` is `0 - 1 = -1`.
    *   Finally, `print({-1})` creates a set containing the value -1 and prints it.

Therefore, the output of the code is `{-1}`.

In [2]:
var=1
def f():
  global var
  var+=1
  def g():
    return var
  return g
a=f()
b=f()
print (a is b)
print(b()>2)
print(a()>2)
print(a!=None)



False
True
True
True


Let's explain the code in `gpFzGcpSrp-A`:

*   `var=1`:
    *   This initializes a global variable `var` with the value `1`.

*   `def f():`:
    *   This defines an outer function `f`.

*   `global var`:
    *   Inside `f`, this statement declares that the `var` being referenced is the global `var`, not a new local variable. This allows `f` to modify the global `var`.

*   `var+=1`:
    *   The global `var` is incremented by 1. Each time `f()` is called, `var` increases.

*   `def g(): return var`:
    *   This defines a nested function `g` inside `f`.
    *   `g` does not have its own `var` and `var` is not explicitly declared as `global` within `g`. In Python, nested functions can access variables from their enclosing scope (the scope of `f` in this case). When `g` is called, it will access the `var` that `f` has access to, which is the global `var`.

*   `return g`:
    *   The function `f` returns the nested function `g` itself, not the result of `g()`.

*   `a=f()`:
    *   `f()` is called for the first time.
    *   `global var` ensures `var` refers to the global variable.
    *   `var` becomes `1 + 1 = 2`.
    *   `f` returns the function `g`. So, `a` now holds a reference to the `g` function defined during this call to `f`.

*   `b=f()`:
    *   `f()` is called for the second time.
    *   `global var` again refers to the global variable.
    *   `var` (which was `2`) becomes `2 + 1 = 3`.
    *   `f` returns the function `g`. So, `b` now holds a reference to the `g` function defined during this second call to `f`.

*   `print (a is b)`:
    *   This checks if `a` and `b` refer to the *exact same function object* in memory. Since `f()` creates and returns a *new* nested `g` function object each time it's called, `a` and `b` will be different function objects. This prints `False`.

*   `print(b()>2)`:
    *   `b()` calls the `g` function that `b` references. This `g` function returns the current value of the global `var`, which is `3`.
    *   `3 > 2` is `True`. This prints `True`.

*   `print(a()>2)`:
    *   `a()` calls the `g` function that `a` references. This `g` function also returns the current value of the global `var`, which is still `3` (because `g` refers to the global `var`, which was last incremented by `b=f()`).
    *   `3 > 2` is `True`. This prints `True`.

*   `print(a!=None)`:
    *   This checks if `a` holds a value that is not `None`. Since `a` holds a function object, it is not `None`. This prints `True`.

In summary, the key takeaway is that `a` and `b` are references to different function objects, but both of these function objects (`g` in this case) 'close over' and share the *same single global variable* `var` from their enclosing scope.

In [3]:
myli= range(-2,2)
m=list(filter(lambda x: True if abs(x)<1 else False,myli))
print(m)
print(len(m))

[0]
1


Let's explain the code in `uq7ZP-aPwQKo`:

*   `myli = range(-2, 2)`:
    *   This line creates a `range` object named `myli`. A `range(start, stop)` object generates a sequence of numbers starting from `start` up to, but not including, `stop`.
    *   So, `myli` will represent the sequence `(-2, -1, 0, 1)`.

*   `m = list(filter(lambda x: True if abs(x)<1 else False, myli))`:
    *   This is the core of the operation. It uses the `filter()` function.
    *   `filter(function, iterable)` constructs an iterator from elements of `iterable` for which `function` returns true.
    *   `lambda x: True if abs(x)<1 else False` is an anonymous function (a lambda function) that serves as the `function` for `filter`.
        *   It takes one argument `x`.
        *   `abs(x)` calculates the absolute value of `x`.
        *   `abs(x) < 1` checks if the absolute value of `x` is less than 1.
        *   If `abs(x) < 1` is `True`, the lambda returns `True`.
        *   Otherwise (`abs(x) < 1` is `False`), the lambda returns `False`.
    *   Let's trace it with `myli`:
        *   For `x = -2`, `abs(-2)` is `2`. `2 < 1` is `False`. Element is filtered out.
        *   For `x = -1`, `abs(-1)` is `1`. `1 < 1` is `False`. Element is filtered out.
        *   For `x = 0`, `abs(0)` is `0`. `0 < 1` is `True`. Element is kept.
        *   For `x = 1`, `abs(1)` is `1`. `1 < 1` is `False`. Element is filtered out.
    *   The `filter` function will yield only `0`.
    *   `list(...)` converts the iterator returned by `filter` into a list.
    *   Therefore, `m` will be `[0]`.

*   `print(m)`:
    *   This prints the list `m`, which is `[0]`.

*   `print(len(m))`:
    *   This prints the number of elements in list `m`. Since `m` is `[0]`, its length is `1`.

So, the output of the code will be:
```
[0]
1
```

In [5]:
my_list=[i for i in range (5,0,-1)]
m=[my_list[i] for i in range(5) if my_list[i]%2==0]
print (m)



[4, 2]


In [11]:
import math
print(math.hypot(3,4)==math.sqrt(25))
print(math.hypot(2,5)==math.trunc(2.5))
#print(math.hypot(2,5)==math.true(2,5))
#print(math.hypot(2,5)==math.truec(2,5))
print(math.ceil(2.5)==math.floor(2.5))

True
False
False


In [5]:
string ='SKY'[::-1]
string=string[-1]
print(string is None)
print(string[0]==string[-1])
print(string[0]=='Y')
print(len(string)==1)



False
True
False
True


In [9]:
class MyClass:
  def __init__(self,initial):
    self.store=initial
  def put(self,new):
    self.store.append(new)
  def get(self):
    return self.store
  def dup(self):
    #self.store.append(self.store[-1])
    #self.put(self.get()[-1])
    self.put(self.store[1])

Object = MyClass([0])    # self.store = [0]
Object.put(1)            # self.store.append(1) → [0, 1]
Object.dup()             # must duplicate last element → [0, 1, 1]
print(Object.get())      # must output [0, 1, 1]



[0, 1, 1]


In [11]:
x = 3 % 1  # Initializes x. 3 % 1 is 0.
y = -1 if x > 0 else 0  # Corrected conditional assignment: y will be 0 as x (0) is not greater than 0.
print(y)

0


In [12]:
def foo(x,y):
  return y(x)+(x+1)
print(foo(1,lambda x: x*x))


3


Let's explain the code in `cufJXoxWnnA2`:

*   `def foo(x, y): return y(x) + (x + 1)`:
    *   This defines a function named `foo` that takes two arguments: `x` and `y`.
    *   Similar to a previous example, `y` is expected to be a function here. Inside `foo`, `y` is called with `x` as its argument (`y(x)`).
    *   The function then returns the sum of the result of `y(x)` and `(x + 1)`.

*   `lambda x: x * x`:
    *   This is an anonymous function (a lambda function) that takes one argument, named `x`.
    *   It returns the square of its input `x` (`x * x`).

*   `print(foo(1, lambda x: x * x))`:
    *   Here, `foo` is called with two arguments:
        1.  `x`: The integer `1`.
        2.  `y`: The lambda function `lambda x: x * x`.
    *   Let's trace `foo`'s execution with these arguments:
        *   `y(x)` becomes `(lambda x: x * x)(1)`, which evaluates to `1 * 1 = 1`.
        *   `(x + 1)` becomes `(1 + 1) = 2`.
        *   The return value of `foo` is `1 + 2 = 3`.
    *   Finally, `print(3)` prints the result to the console.

Therefore, the output of the code is `3`.

In [13]:
def f(x,y):
  num,denom=x,y
  def g():
    return num/denom
  return g
a=f(1,2)
b=f(3,4)
print(a()==4)
print(a is not None)
print(b()==4)
print(a!=b)



False
True
False
True


Let's explain the code in `WbjgGAZSsCgo`:

*   `def f(x,y):`:
    *   This defines an outer function `f` that takes two arguments: `x` and `y`.

*   `num,denom=x,y`:
    *   Inside `f`, `x` and `y` are assigned to local variables `num` and `denom`. These variables will be 'captured' by the nested function `g`.

*   `def g(): return num/denom`:
    *   This defines a nested function `g` inside `f`.
    *   `g` does not take any arguments itself but has access to `num` and `denom` from its enclosing function `f`'s scope. This is a classic example of a **closure**.
    *   When `g` is called, it will return the result of `num` divided by `denom` using the values captured at the time `f` was called.

*   `return g`:
    *   The function `f` returns the nested function `g` itself, not the result of calling `g()`.

*   `a=f(1,2)`:
    *   `f` is called with `x=1` and `y=2`.
    *   Inside this call, `num` becomes `1` and `denom` becomes `2`.
    *   `f` returns a version of `g` where `num=1` and `denom=2`. This returned function is assigned to `a`.

*   `b=f(3,4)`:
    *   `f` is called with `x=3` and `y=4`.
    *   Inside this call, `num` becomes `3` and `denom` becomes `4`.
    *   `f` returns *another* version of `g` (a distinct function object) where `num=3` and `denom=4`. This returned function is assigned to `b`.

*   `print(a()==4)`:
    *   `a()` executes the `g` function associated with `a`. It calculates `1 / 2 = 0.5`.
    *   The comparison `0.5 == 4` is `False`. So, `False` is printed.

*   `print(a is not None)`:
    *   This checks if `a` holds a value that is not `None`. Since `a` holds a function object, it is not `None`. So, `True` is printed.

*   `print(b()==4)`:
    *   `b()` executes the `g` function associated with `b`. It calculates `3 / 4 = 0.75`.
    *   The comparison `0.75 == 4` is `False`. So, `False` is printed.

*   `print(a!=b)`:
    *   This checks if `a` and `b` are references to different function objects. Since `f` returns a *new* `g` function object each time it's called, `a` and `b` refer to different function instances. So, `True` is printed.

Therefore, the output of the code is:
```
False
True
False
True
```

In [14]:
i=0
while i!=0:
  i=i-1
else:
  i=i+1
print(i)




1


In [15]:
for i in range(10):
  print (i)

0
1
2
3
4
5
6
7
8
9


In [25]:
i=5
while i>0:
  #print('in while loop '+ str(i))
  i=i//2
  print('in integer division '+ str(i))
  if i%2==0:
    #print('in modulus division '+ str(i%2))
    break
  else:
    i+=1
    #print('in else statement '+ str(i))
else:
  print(i)


in while loop 5
in integer division 2


Let's explain the code in `_J02vyGu3rA4`:

*   `i = 5`:
    *   Initializes the variable `i` with the value `5`.

*   `while i > 0:`:
    *   This is a `while` loop that continues as long as the condition `i > 0` is `True`.

*   **Loop Iteration 1:**
    *   `i` is `5`. Condition `5 > 0` is `True`. Enter the loop.
    *   `i = i // 2`: Integer division. `5 // 2` is `2`. So, `i` becomes `2`.
    *   `print('in integer division ' + str(i))`: Prints `in integer division 2`.
    *   `if i % 2 == 0:`:
        *   `i` is `2`. `2 % 2` is `0`. Condition `0 == 0` is `True`.
    *   `break`:
        *   The `break` statement is executed, which immediately terminates the `while` loop. The `else` block associated with the `while` loop is **not** executed when the loop is terminated by a `break`.

*   **After the loop:**
    *   The code continues after the `while...else` block.
    *   There is no `print(i)` statement outside the `while` or its `else` block in the provided code, but the output indicates that the `print('in while loop ' + str(i))` (which was commented out) must have been uncommented and executed before the `break` for the observed output `in while loop 5` and `in integer division 2` to occur. However, based strictly on the provided *latest notebook content*, the only print statement inside the loop is `'in integer division ' + str(i)`.
    *   If the commented lines were active:
        *   `print('in while loop ' + str(i))` would print `in while loop 5`.
        *   `i` becomes `2`.
        *   `print('in integer division ' + str(i))` would print `in integer division 2`.
        *   Then `break`.
    *   The final state of `i` after the execution of the provided code block (ignoring the commented `print` statements) would be `2`.

In [26]:
for i in range(1,3):
  print("*",end="")
else:
  print("*")

***


In [2]:
a=[1]
b=a
a[0]=0

print (len(a)==len(b))

print(b[0] == a[0] )
#print(a[0]==b[0])
print ((a[0])+1==b[0])

True
True
True
False


In [2]:
lst=[1,2,3,4]
lst=lst[-3:-2]
print (lst)
lst=lst[-1]
print(lst)

[2]
2


In [4]:
original_str = 'abcd'
s_list = []
for i in range(len(original_str)):
  s_list.append(original_str[i].upper())
s = "".join(s_list)
print(s)

ABCD


In [5]:
list1 = [False for i in range (1, 10) ]
list2 = list1 [-1:1:-1]
print(list1)
print(list2)


[False, False, False, False, False, False, False, False, False]
[False, False, False, False, False, False, False]
